geograph/geograph.py

```diff
@@ -528,6 +528,28 @@ def merge_classes(
         if not set(class_list).issubset(self.df["class_label"].unique()):
             raise ValueError("`class_list` must only contain valid class names.")
         self.df.loc[self.df["class_label"].isin(class_list), "class_label"] = new_name
+        # Get set of indices of nodes for the new class
+        node_set = set(
+            self.df.loc[
+                self.df["class_label"].isin(class_list), "class_label"
+            ].index.tolist()
+        )
+        merged_neighbours = set()
+        for node in node_set:
+            # Skip iteration if node has already been merged
+            if node in merged_neighbours:
+                continue
```

Comment:
I think you're right, 3 and 13 won't be merged here. However, that doesn't actually happen because of `merged_neighbours`, since the nodes which have indexes which survive to the end (the "primary" node in each loop iteration), like 3 and 13 in this example, are not put into the `merged_neighbours` set. Instead it happens because the merged nodes are assigned indices in `node_set` and are not returned to in the main loop. It also happens because of the order of iteration - it wouldn't happen if node 12 was the "primary" node instead of node 13 in step 3.2. And because we're iterating over a set the order of iteration is random too :D

I propose this solution:
```python
node_set = set(self.df.loc[self.df["class_label"].isin(class_list), "class_label"].index)
merged_neighbours = set()
while True:
    num_merges = 0
    for node in node_set:
        # Skip iteration if node has already been merged
        if node in merged_neighbours:
            continue
        nodes_to_merge = []
        # Get list of neighbours of node, and append them to the merge list
        # if they have the label of the new class
        neighbours = self.graph[node]
        for neighbour in neighbours:
            if neighbour in node_set:
                nodes_to_merge.append(neighbour)
        # Merge nodes with neighbours where both have the label of the new
        # class. Crucially, we assign the merged node an index in `node_set`.
        self.merge_nodes(nodes_to_merge, class_label=new_name, final_index=node)
        merged_neighbours.update(nodes_to_merge)
        num_merges += 1
    if num_merges == 0:
        break
```

Reply:
Nice - I agree with your analysis, thank you for clarifying!

Regarding the proposed solution:
Yes I think that works! (: 
How intensive is the operation? Does the sequential merging take much time? If so, we might be able to optimize it by leveraging the graph structure (forgive the pseudo code):

> (1) Creating a set of all nodes with this class label (I'll call it `_node_set`, `_` indicating that it's temporary )
> (2) While `_node_set` is non-empty:
>    (2.0) Pop the first node in `current_node = _node_set.pop()` 
>    (2.1) From `current_node`, and do a graph traversal (e.g. bfs) along the nearest neighbors to find all nodes that can be reached from `current_node` by going through nodes with the same class label. Add those nodes to a list called `nodes_to_merge`
>    (2.2) Merge all nodes in `nodes_to_merge`, with the final index of `current_node` 
>    (2.2) Remove `nodes_to_merge` from `_node_set`

Reply:
Yeah I thought of this but was too lazy to implement it - I did some testing now and the current implementation is very slow, so I'll make an issue to add this in another PR and put a warning in the comment of this method.

Reply:
Creating an issue for it sounds good to me - Thank you (: